### ISSUE TITLE
torch.autograd.Function unexpectedly modifies gradients for unused parameters

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#58511]
Not sure if this is a bug or expected behavior, but opening the issue for discussion/clarification. In the following repro, `b` is not used for loss computation and therefore I'd expect the gradient for the linear layer `self.b` to stay as `None`, however, the gradient is `0`:

```
import torch
import torch.nn as nn
from torch.autograd import Function


class PassThrough(Function):
    @staticmethod
    def forward(ctx, *inputs):
        return inputs

    @staticmethod
    def backward(ctx, *grad_outputs):
        print(f"grad_outputs {grad_outputs}")
        return grad_outputs


class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Linear(1, 1, bias=False)
        self.b = nn.Linear(1, 1, bias=False)

    def forward(self, x):
        a, b = self.a(x), self.b(x)
        ret = a, b
        new_a, new_b = PassThrough.apply(a, b)
        ret = new_a, new_b
        return ret

model = MyModel()

def print_grads():
    for param_name, param in model.named_parameters():
        print(f"{param_name} : {param.grad}")


inp = torch.ones(1)
print("-- before backward ---")
model.zero_grad()
print_grads()

model.zero_grad()
out = model(inp)
loss = out[0].sum()
print("Calling backward...")
loss.backward()
print("-- after bwd --")
print_grads()
```

This produces the output:
```
-- before backward ---
a.weight : None
b.weight : None
Calling backward...
grad_outputs (tensor([1.]), tensor([0.]))
-- after bwd --
a.weight : tensor([[1.]])
b.weight : tensor([[0.]])
```

It looks like since `out[1]` is not used in computing loss, the `torch.autograd.Function` fills in a `0` for the corresponding grad tensor in `grad_outputs`, and as a result, the .grad field gets set to `0`.  However, this particular `torch.autograd.Function` as defined doesn't really do anything (it just forwards the inputs/outputs) so I would expect it the resulting gradients to not be different than if the `torch.autograd.Function` was not used. When the `torch.autograd.Function` usage is omitted from the forward pass, the gradient for `b` is not touched as expected:

```
-- before backward ---
a.weight : None
b.weight : None
Calling backward...
-- after bwd --
a.weight : tensor([[1.]])
b.weight : None
```

Ideally I would expect it to be the same even with the torch.autograd.Function, since the gradients are not modified at all. Would appreciate any comments/thoughts on this, thank you!

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

**Comment by rohan-varma** [comment]
Support for this actually exists with https://pytorch.org/docs/stable/autograd.html#torch.autograd.function._ContextMethodMixin.set_materialize_grads, closing.
